### PR TITLE
Optimize batchnorm1d using 2D kernel

### PR DESCRIPTION
**Pull request by EsdeathYZH**
### PR types

Function optimization
### PR changes

OPs
### Describe

[调研设计文档](https://github.com/PaddlePaddle/community/pull/135)

1 Motivation：
+ BatchNorm1D算子与pytorch相比性能差

2 Design：
+ 根据Batch size判断是否使用cudnn库进行计算，当batch size大于阈值时使用native kernel
+ 使用2D tile的方式划分block，提高访存局部性以及增加资源利用率

3 Evaluation：
测试环境：NVIDIA V100 GPU

|  输入形状   | Pytorch | Oneflow | Paddle优化前  | Paddle优化后
|  ----  | ----  | ----  | ----  | ----  |
| [126000, 16]  | 0.014 | 0.082 | 0.773 | 0.023 |
| [136000, 32]  | 0.012 | 0.217 | Error | 0.022 |
| [215857, 32]  | 0.015 | 0.371 | Error | 0.022 |
| [215857, 64]  | 0.025 | 0.900 | Error | 0.029 |
| [143042, 64]  | 0.020 | 0.544 | Error | 0.020 |
| [62929, 64]  | 0.012 | 0.219 | 0.419 | 0.020 |
| [136000, 16, 16]  | 0.961 | 0.967 | 0.971 | 0.092 |
| [136000, 32, 32]  | 1.015 | 1.019 | 1.023| 0.312 |
| [1000000, 16, 16]  | 3.956 | 7.153 | 7.081 | 0.866 |
| [1000000, 32, 32]  | 4.859 | 7.487 | 7.42 | 2.457 |

[N, C]输入获得了33倍的性能提升，[N, C, L]输入获得了8.2倍的性能提升

### CODE REVIEW

File: paddle/phi/kernels/gpu/batch_norm_kernel.cu [hunk 1]
```diff
@@ -137,6 +138,371 @@ static __global__ LAUNCH_BOUNDS(BlockDim) void BNForwardTraining(
...
+    // vertical block sum
+    int tid = threadIdx.x + threadIdx.y * blockDim.x;
+#pragma unroll
+    for (int offset = blockDim.y / 2; offset > 0; offset >>= 1) {
+      if (threadIdx.y < offset * 2) {
+        smem_sum[tid] = x_sum;
+        smem_square_sum[tid] = x_square_sum;
+      }
+      __syncthreads();
+      if (threadIdx.y < offset && threadIdx.y + offset < blockDim.y) {
```
Comment (zkh2016) [inline paddle/phi/kernels/gpu/batch_norm_kernel.cu]: 后面这个条件是不是多余了？

File: paddle/phi/kernels/gpu/batch_norm_kernel.cu [hunk 2]
```diff
@@ -137,6 +138,371 @@ static __global__ LAUNCH_BOUNDS(BlockDim) void BNForwardTraining(
...
+      if (is_last_block_done) {
+        x_sum = static_cast<BatchNormParamType<T>>(0);
+        x_square_sum = static_cast<BatchNormParamType<T>>(0);
+        // thread sum
+        for (int y = threadIdx.y; y < gridDim.y; y += blockDim.y) {
+          x_sum += staging_sum[i + y * C];
+          x_square_sum += staging_square_sum[i + y * C];
+        }
+
+        // vertical block sum
```
Comment (zkh2016) [inline paddle/phi/kernels/gpu/batch_norm_kernel.cu:278]: 这里和183行的处理是一样的， 可以提取出来。

File: paddle/phi/kernels/gpu/batch_norm_kernel.cu (hunk 1 above)
Comment (EsdeathYZH) [reply paddle/phi/kernels/gpu/batch_norm_kernel.cu]: 已去掉

File: paddle/phi/kernels/gpu/batch_norm_kernel.cu (hunk 2 above)
Comment (EsdeathYZH) [reply paddle/phi/kernels/gpu/batch_norm_kernel.cu:278]: done

**Review by zkh2016 — APPROVED** [review]

File: paddle/phi/kernels/gpu/batch_norm_grad_kernel.cu [hunk 3]
```diff
@@ -591,10 +591,12 @@ void BatchNormGradRawKernel(const Context &ctx,
 //             ctx.GetPlace()),
 //         epsilon, saved_mean_data, saved_var_data));
 #else
-      // CUDNN PER_ACTIVATION mode only support small batch size
+      // CUDNN only support small batch size
```
Comment (xingfeng01) [inline paddle/phi/kernels/gpu/batch_norm_grad_kernel.cu:594]: 和版本有关吗 ？

**Review by AnnaTrainingG — APPROVED** [review]

File: paddle/phi/kernels/gpu/batch_norm_kernel.cu [hunk 4]
```diff
@@ -137,6 +138,398 @@ static __global__ LAUNCH_BOUNDS(BlockDim) void BNForwardTraining(
   }
 }
 
+template <typename T>
```
Comment (xingfeng01) [inline paddle/phi/kernels/gpu/batch_norm_kernel.cu:141]: 后续补充注释

File: paddle/phi/kernels/gpu/batch_norm_grad_kernel.cu (hunk 3 above)
Comment (zkh2016) [reply paddle/phi/kernels/gpu/batch_norm_grad_kernel.cu:594]: 目前用到的版本是不支持太大的batch_size。
